File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/iteration/144. Binary Tree Preorder Traversal.cpp
// 144. Binary Tree Preorder Traversal
// https://leetcode.com/problems/binary-tree-preorder-traversal/

// clang++ 144.\ Binary\ Tree\ Preorder\ Traversal.cpp -std=c++17

#include "../../../printUtils.hpp"
#include <stack>
#include <vector>

template <typename T>
std::vector<T> preOrderTraversal(TreeNode<T>* root) {
    std::vector<T> result;
    std::stack<TreeNode<T>*> stk;

    while (root != nullptr || !stk.empty()) {
        if (root != nullptr) {
            result.push_back(root->val);
            stk.push(root->right);
            root = root->left;
        } else {
            root = stk.top();
            stk.pop();
        }
    }

    return result;
}

int main() {
    TreeNode<int>* root = buildTree({1, 2, 3, 4, -1, 5, 6}, -1);
    std::cout << "Preorder traversal: ";
    printArray(preOrderTraversal(root));
    deleteTree(root);
    
    return 0;
}

File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/iteration/144. Binary Tree Preorder Traversal.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def inorderTraversal(root):
    if not root:
        return []
    
    result = []
    result.append(root.val)
    result.extend(inorderTraversal(root.left))
    result.extend(inorderTraversal(root.right))
    return result

# 测试用例
root = TreeNode(1, None, TreeNode(2, TreeNode(3)))
print(inorderTraversal(root)) 
